src/util.rs

```diff
@@ -0,0 +1,76 @@
+use core::ops::{Deref, DerefMut};
+
+pub enum MaybeRef<'a, T> {
+    Ref(&'a T),
+    Own(T),
+}
```

Comment:
So I actually thought of a way we can remove the need for this last night by slightly adjusting iter parsers. I think this is probably not needed now. Other than that, this PR looks good.